### ISSUE TITLE
NaNs in fiberassign file coordinates

### ISSUE THREAD

**Issue by djschlegel** [desihub/fiberassign#323]
All fiberassign files since starting to use the real focal plane models on 2021-03-17 have 14-15 rows with NaN values or FIBERASSIGN_X,FIBERASSIGN_Y,TARGET_RA,TARGET_DEC.  This is unexpected behavior, and turns out to cause problems with the ICS split-exposure code as implemented.

The 14 fibers that always have NaN values are: 367 1577 1763 1787 1868 2132 3094 3520 3587 3634 3668        3788 4420 4625.  Fiber 1288 sometimes has NaNs.  Note these are all positioners labeled as “unused” in the focal plane model, although it's only a small subset of those positioners.

This includes all fiberassign files checked in here:
https://desi.lbl.gov/trac/browser/data/tiles/trunk/000

**Comment by tskisner** [comment]
Thanks @djschlegel.  I'll look at those positioners and see if there is anything they have in common.  The "unassigned" bit (bit 1) of FIBERSTATUS is set by fiberassign and is not part of the focal plane model.  If these were unassigned due to being non-functional then that would point towards a problem in the code here:

https://github.com/desihub/fiberassign/blob/907ce0d7f62b6cda31f0856f4b2207945244dca4/py/fiberassign/assign.py#L329

and if these are functional, but simply unassigned, then it points to a problem here:

https://github.com/desihub/fiberassign/blob/907ce0d7f62b6cda31f0856f4b2207945244dca4/py/fiberassign/assign.py#L350

Either way, will sort it out.

**Comment by tskisner** [comment]
Just another note on this, when plotting positioners recently for #325, I noticed that `fba_plot_results` printed some warnings about assigned RA / DEC values that were not reachable.  This may be related or not.  Adding a note here as a clue.

**Comment by araichoor** [comment]
I think this is on Ted s plate; from my understanding (ie reading Ted s comment on this ticket), this has not been fixed.
Ted, please let me know if I can help on this one.

**Comment by dstndstn** [comment]
I was digging into another issue (#325) and stumbled upon this.

The proximal cause seems to be that hw.thetaphi_to_xy is returning None,None for the X,Y positions, which turn into NaNs, which then get converted into RA,Dec NaNs as well.  This can happen for UNASSIGNED as well as STUCK/BROKEN positioners, and occurs because theta or phi are outside the min/max ranges.

**Comment by tskisner** [comment]
Thanks for digging on that- strange since the theta / phi angles passed to that function for such positioners are either the fixed angles from the DB (for the case of stuck / broken) or should be the "parking" angles for unassigned positioners, which (I thought) were chosen to always be in the allowable range...

**Comment by dstndstn** [comment]
Here's one example (in fact, one of the ones in the center of a hex that is bad in #325...):

thetaphi_to_xy failed: out of range:
phi bad: phi 3.1765, phi_0 -0.126728, min 2.1164, max 3.1765
diff_hi 3.30323, diff_lo 2.97996
Loc 432 : STUCK or BROKEN fiber; X,Y None None
  pos args: (43.38153393173591, -350.8650990786089) -3.380494714968659 3.1764992386296798 3.227534055709839 3.2244186401367188 -1.2407100265366984 -0.1267282681708859 -3.380494714968659 2.11640359879966
46 3.3281348374088293 3.1764992386296798

**Comment by dstndstn** [comment]
So maybe the fixed PHI values for these BROKEN positioners are set to MAX rather than MAX-PHI_0 ?

**Comment by tskisner** [comment]
Looking at comments in the DB dump files...
```
# - {name: POS_T, unit: deg, datatype: float64, description: 'current value of internally-tracked theta angle, a.k.a. "posintT" or t_int'}
# - {name: POS_P, unit: deg, datatype: float64, description: 'current value of internally-tracked phi angle, a.k.a. "posintP" or p_int'}

<snip>

# - {name: MAX_T, unit: deg, datatype: float64, description: max targetable internally-tracked theta angle (i.e. "POS_T" or "posintT"
#     or "t_int"}
# - {name: MIN_T, unit: deg, datatype: float64, description: min targetable internally-tracked theta angle (i.e. "POS_T" or "posintT"
#     or "t_int"}
# - {name: MAX_P, unit: deg, datatype: float64, description: max targetable internally-tracked phi angle (i.e. "POS_P" or "posintP" or
#     "p_int"}
# - {name: MIN_P, unit: deg, datatype: float64, description: min targetable internally-tracked phi angle (i.e. "POS_P" or "posintP" or
#     "p_int"}
```
So it seems that `POS_T` / `POS_P` are relative to the same starting angle as the `MIN_*` / `MAX_*` values.  When testing for valid theta / phi values:

https://github.com/desihub/fiberassign/blob/41e0aa67593e511acec6a18060186acbd92d954a/src/hardware.cpp#L567

I look at the difference of the angle compared to `OFFSET_T` / `OFFSET_P` (i.e. theta / phi zero point).  Then compare with the min / max angle which are assumed to also be relative to these OFFSET / zero points.

**Comment by tskisner** [comment]
Also, note that for restricted positioners, the `eo_phi` angle in the dump is relative to the coordinate frame along the theta arm.  So to get the restricted positioner `MIN_P` angle, the offset is subtracted when syncing:

https://github.com/desihub/desimodel/blob/8a36c7d83c2421a50d24cc51823bd474578cd768/py/desimodel/inputs/focalplane_sync.py#L317

**Comment by dstndstn** [comment]
Wait a minute.

Your reading is that MIN_T <= POS_T <= MAX_T, but that POS_T should be offset by OFFSET_T when doing geometry on the focal plane, yes?

Then shouldn't outside_theta_phi_range be just that *trivial* comparison?

And by the way, in the 2021-03-17 database dump, 24 positioners have POS_T and POS_P values that fail that criterion!  (23 of them are either NONFUNCTIONAL or not(FIBER_INTACT)).

**Comment by tskisner** [comment]
From a slack chat with Kevin:

> I looked at a few of the numbers and they do have a simple explaination: we give you min/max t/p based on the targetable range (the range they are allowed to move in). There's a few degrees difference between that and the full range (which is permitted for TP updates from measurements) since hardware problems or measurement errors can push you outside of where you wanted to go. This applies to all of the theta cases I see here.

Ongoing discussion about whether DB dump should include both the max and targetable ranges, or whether we should just treat POS T/P values as always valid.

**Comment by dstndstn** [comment]
This was fixed as part of #340 -- if a STUCK/BROKEN positioner had a THETA/PHI outside its MIN/MAX range (which can happen), then its coordinate transformations bailed out, and we ended up with NaNs in the output file.

**Comment by araichoor** [comment]
I confirm that with the large runs I did with the stuck+sky branch, there are no more NaN in TARGET_RA, TARGET_DEC in the fiberassign-TILEID.fits.gz.
I checked the ASSIGN extension for all tiles from a SGC run.
perfect.
